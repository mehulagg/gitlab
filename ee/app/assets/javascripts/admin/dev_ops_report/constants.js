import { s__, __ } from '~/locale';

export const MAX_REQUEST_COUNT = 10;

export const DEVOPS_ADOPTION_SEGMENT_MODAL_ID = 'devopsSegmentModal';

export const DEVOPS_ADOPTION_SEGMENT_DELETE_MODAL_ID = 'devopsSegmentDeleteModal';

export const DATE_TIME_FORMAT = 'yyyy-mm-dd HH:MM';

export const DEVOPS_ADOPTION_ERROR_KEYS = {
  groups: 'groupsError',
  segments: 'segmentsError',
};

export const DEVOPS_ADOPTION_STRINGS = {
  app: {
    [DEVOPS_ADOPTION_ERROR_KEYS.groups]: s__(
      'DevopsAdoption|There was an error fetching Groups. Please refresh the page to try again.',
    ),
    [DEVOPS_ADOPTION_ERROR_KEYS.segments]: s__(
      'DevopsAdoption|There was an error fetching Segments. Please refresh the page to try again.',
    ),
    tableHeader: {
      text: s__(
        'DevopsAdoption|Feature adoption is based on usage in the last calendar month. Last updated: %{timestamp}.',
      ),
      button: s__('DevopsAdoption|Add new segment'),
    },
  },
  emptyState: {
    title: s__('DevopsAdoption|Add a segment to get started'),
    description: s__(
      'DevopsAdoption|DevOps adoption uses segments to track adoption across key features. Segments are a way to track multiple related projects and groups at once. For example, you could create a segment for the engineering department or a particular product team.',
    ),
    button: s__('DevopsAdoption|Add new segment'),
  },
  modal: {
    addingTitle: s__('DevopsAdoption|New segment'),
    editingTitle: s__('DevopsAdoption|Edit segment'),
    addingButton: s__('DevopsAdoption|Create new segment'),
    editingButton: s__('DevopsAdoption|Save changes'),
    cancel: __('Cancel'),
    namePlaceholder: s__('DevopsAdoption|My segment'),
    filterPlaceholder: s__('DevopsAdoption|Filter by name'),
    nameLabel: s__('DevopsAdoption|Name'),
    selectedGroupsTextSingular: s__('DevopsAdoption|%{selectedCount} group selected (20 max)'),
    selectedGroupsTextPlural: s__('DevopsAdoption|%{selectedCount} groups selected (20 max)'),
    error: s__('DevopsAdoption|An error occured while saving the segment. Please try again.'),
    noResults: s__('DevopsAdoption|No filter results.'),
  },
  table: {
    editButton: s__('DevopsAdoption|Edit segment'),
    deleteButton: s__('DevopsAdoption|Delete segment'),
    headers: {
      name: {
        label: s__('DevopsAdoption|Segment'),
      },
      issueOpened: {
        label: s__('DevopsAdoption|Issues'),
        tooltip: s__('DevopsAdoption|At least 1 issue opened'),
      },
      mergeRequestOpened: {
        label: s__('DevopsAdoption|MRs'),
        tooltip: s__('DevopsAdoption|At least 1 MR opened'),
      },
      mergeRequestApproved: {
        label: s__('DevopsAdoption|Approvals'),
        tooltip: s__('DevopsAdoption|At least 1 approval on an MR'),
      },
      runnerConfigured: {
        label: s__('DevopsAdoption|Runners'),
        tooltip: s__('DevopsAdoption|Runner configured for project/group'),
      },
      pipelineSucceeded: {
        label: s__('DevopsAdoption|Pipelines'),
        tooltip: s__('DevopsAdoption|At least 1 pipeline successfully run'),
      },
      deploySucceeded: {
        label: s__('DevopsAdoption|Deploys'),
        tooltip: s__('DevopsAdoption|At least 1 deploy'),
      },
      securityScanSucceeded: {
        label: s__('DevopsAdoption|Scanning'),
        tooltip: s__('DevopsAdoption|At least 1 security scan of any type run in pipeline'),
      },
    },
  },
  deleteModal: {
    title: s__('DevopsAdoption|Confirm delete segment'),
    confirmationMessage: s__('DevopsAdoption|Are you sure that you would like to delete %{name}?'),
    cancel: __('Cancel'),
    confirm: s__('DevopsAdoption|Delete segment'),
    error: s__('DevopsAdoption|An error occured while deleting the segment. Please try again.'),
  },
  tableCell: {
    trueText: s__('DevopsAdoption|Adopted'),
    falseText: s__('DevopsAdoption|Not adopted'),
  },
};

export const DEVOPS_ADOPTION_TABLE_TEST_IDS = {
  TABLE_HEADERS: 'header',
  SEGMENT: 'segmentCol',
  ISSUES: 'issuesCol',
  MRS: 'mrsCol',
  APPROVALS: 'approvalsCol',
  RUNNERS: 'runnersCol',
  PIPELINES: 'pipelinesCol',
  DEPLOYS: 'deploysCol',
  ACTIONS: 'actionsCol',
  SCANNING: 'scanningCol',
};
