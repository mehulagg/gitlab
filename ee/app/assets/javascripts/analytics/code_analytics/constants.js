export const PROJECTS_PER_PAGE = 50;
export const FILE_QUANTITIES = [25, 50, 100, 250, 500];
export const DEFAULT_FILE_QUANTITY = FILE_QUANTITIES[2];
export const DEFAULT_DAYS_IN_PAST = 30;
export const DEFAULT_SPACING = 5;
