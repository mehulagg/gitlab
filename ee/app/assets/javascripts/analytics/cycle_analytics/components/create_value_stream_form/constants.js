import { s__ } from '~/locale';

export const I18N = {
  SELECT_START_EVENT: s__('CustomCycleAnalytics|Select start event'),
  SELECT_END_EVENT: s__('CustomCycleAnalytics|Select stop event'),
};

export const ERRORS = {
  START_EVENT_REQUIRED: s__('CustomCycleAnalytics|Please select a start event first'),
  STAGE_NAME_EXISTS: s__('CustomCycleAnalytics|Stage name already exists'),
  INVALID_EVENT_PAIRS: s__(
    'CustomCycleAnalytics|Start event changed, please select a valid end event',
  ),
};

export const defaultErrors = {
  id: [],
  name: [],
  startEventIdentifier: [],
  startEventLabelId: [],
  endEventIdentifier: [],
  endEventLabelId: [],
};

export const defaultFields = {
  id: null,
  name: null,
  startEventIdentifier: null,
  startEventLabelId: null,
  endEventIdentifier: null,
  endEventLabelId: null,
};
