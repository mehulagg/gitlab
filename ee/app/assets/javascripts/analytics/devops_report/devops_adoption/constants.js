import { s__, __, sprintf } from '~/locale';

export const DEFAULT_POLLING_INTERVAL = 30000;

export const MAX_SEGMENTS = 30;

export const MAX_REQUEST_COUNT = 10;

export const DEVOPS_ADOPTION_SEGMENT_MODAL_ID = 'devopsSegmentModal';

export const DEVOPS_ADOPTION_SEGMENT_DELETE_MODAL_ID = 'devopsSegmentDeleteModal';

export const DATE_TIME_FORMAT = 'yyyy-mm-dd HH:MM';

export const DEVOPS_ADOPTION_ERROR_KEYS = {
  groups: 'groupsError',
  segments: 'segmentsError',
  addSegment: 'addSegmentsError',
};

export const DEVOPS_ADOPTION_GROUP_LEVEL_LABEL = s__('DevopsAdoption|Add/remove sub-groups');

export const DEVOPS_ADOPTION_TABLE_REMOVE_BUTTON_DISABLED = s__(
  'DevopsAdoption|You cannot remove the group you are currently in.',
);

export const DEVOPS_ADOPTION_STRINGS = {
  app: {
    [DEVOPS_ADOPTION_ERROR_KEYS.groups]: s__(
      'DevopsAdoption|There was an error fetching Groups. Please refresh the page.',
    ),
    [DEVOPS_ADOPTION_ERROR_KEYS.segments]: s__(
      'DevopsAdoption|There was an error fetching Group adoption data. Please refresh the page.',
    ),
    [DEVOPS_ADOPTION_ERROR_KEYS.addSegment]: s__(
      'DevopsAdoption|There was an error enabling the current group. Please refresh the page.',
    ),
    tableHeader: {
      text: s__(
        'DevopsAdoption|Feature adoption is based on usage in the current calendar month. Last updated: %{timestamp}.',
      ),
      button: s__('DevopsAdoption|Add/remove groups'),
      buttonTooltip: sprintf(s__('DevopsAdoption|Maximum %{maxSegments} groups allowed'), {
        maxSegments: MAX_SEGMENTS,
      }),
    },
  },
  emptyState: {
    title: s__('DevopsAdoption|Add a group to get started'),
    description: s__(
      'DevopsAdoption|DevOps adoption tracks the use of key features across your favorite groups. Add a group to the table to begin.',
    ),
    button: s__('DevopsAdoption|Add Group'),
  },
  modal: {
    addingTitle: s__('DevopsAdoption|Add/remove groups'),
    addingButton: s__('DevopsAdoption|Save changes'),
    cancel: __('Cancel'),
    namePlaceholder: s__('DevopsAdoption|My group'),
    filterPlaceholder: s__('DevopsAdoption|Filter by name'),
    error: s__('DevopsAdoption|An error occurred while saving changes. Please try again.'),
    noResults: s__('DevopsAdoption|No filter results.'),
  },
  table: {
    removeButton: s__('DevopsAdoption|Remove Group from the table.'),
  },
  deleteModal: {
    title: s__('DevopsAdoption|Confirm remove Group'),
    confirmationMessage: s__(
      'DevopsAdoption|Are you sure that you would like to remove %{name} from the table?',
    ),
    cancel: __('Cancel'),
    confirm: s__('DevopsAdoption|Remove Group'),
    error: s__('DevopsAdoption|An error occurred while removing the group. Please try again.'),
  },
  tableCell: {
    trueText: s__('DevopsAdoption|Adopted'),
    falseText: s__('DevopsAdoption|Not adopted'),
  },
};

export const DEVOPS_ADOPTION_TABLE_TEST_IDS = {
  TABLE_HEADERS: 'header',
  SEGMENT: 'segmentCol',
  ISSUES: 'issuesCol',
  MRS: 'mrsCol',
  APPROVALS: 'approvalsCol',
  RUNNERS: 'runnersCol',
  PIPELINES: 'pipelinesCol',
  DEPLOYS: 'deploysCol',
  ACTIONS: 'actionsCol',
  SCANNING: 'scanningCol',
  LOCAL_STORAGE_SORT_BY: 'localStorageSortBy',
  LOCAL_STORAGE_SORT_DESC: 'localStorageSortDesc',
};

export const DEVOPS_ADOPTION_SEGMENTS_TABLE_SORT_BY_STORAGE_KEY =
  'devops_adoption_segments_table_sort_by';

export const DEVOPS_ADOPTION_SEGMENTS_TABLE_SORT_DESC_STORAGE_KEY =
  'devops_adoption_segments_table_sort_desc';

export const DEVOPS_ADOPTION_GROUP_COL_LABEL = s__('DevopsAdoption|Group');

export const DEVOPS_ADOPTION_TABLE_CONFIGURATION = [
  {
    title: s__('DevopsAdoption|Adoption'),
    cols: [
      {
        key: 'issueOpened',
        label: s__('DevopsAdoption|Issues'),
        tooltip: s__('DevopsAdoption|At least 1 issue opened'),
      },
      {
        key: 'mergeRequestOpened',
        label: s__('DevopsAdoption|MRs'),
        tooltip: s__('DevopsAdoption|At least 1 MR opened'),
      },
      {
        key: 'mergeRequestApproved',
        label: s__('DevopsAdoption|Approvals'),
        tooltip: s__('DevopsAdoption|At least 1 approval on an MR'),
      },
      {
        key: 'runnerConfigured',
        label: s__('DevopsAdoption|Runners'),
        tooltip: s__('DevopsAdoption|Runner configured for project/group'),
      },
      {
        key: 'pipelineSucceeded',
        label: s__('DevopsAdoption|Pipelines'),
        tooltip: s__('DevopsAdoption|At least 1 pipeline successfully run'),
      },
      {
        key: 'deploySucceeded',
        label: s__('DevopsAdoption|Deploys'),
        tooltip: s__('DevopsAdoption|At least 1 deploy'),
      },
      {
        key: 'securityScanSucceeded',
        label: s__('DevopsAdoption|Scanning'),
        tooltip: s__('DevopsAdoption|At least 1 security scan of any type run in pipeline'),
      },
    ],
  },
];
