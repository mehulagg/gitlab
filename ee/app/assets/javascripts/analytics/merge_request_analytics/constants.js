import { __ } from '~/locale';

export const DEFAULT_NUMBER_OF_DAYS = 365;
export const THROUGHPUT_STRINGS = {
  CHART_TITLE: __('Throughput'),
  Y_AXIS_TITLE: __('Merge Requests merged'),
  X_AXIS_TITLE: __('Month'),
  CHART_DESCRIPTION: __('The number of merge requests merged by month.'),
  NO_DATA: __('There is no data available.'),
};
