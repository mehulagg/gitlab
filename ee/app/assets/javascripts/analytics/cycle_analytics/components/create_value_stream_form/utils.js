import { isStartEvent, getAllowedEndEvents, eventToOption, eventsByIdentifier } from '../../utils';
import { I18N, ERRORS, defaultErrors, defaultFields } from './constants';
import { DEFAULT_STAGE_NAMES } from '../../constants';

export const startEventOptions = eventsList => [
  { value: null, text: I18N.SELECT_START_EVENT },
  ...eventsList.filter(isStartEvent).map(eventToOption),
];

export const endEventOptions = (eventsList, startEventIdentifier) => {
  const endEvents = getAllowedEndEvents(eventsList, startEventIdentifier);
  return [
    { value: null, text: I18N.SELECT_END_EVENT },
    ...eventsByIdentifier(eventsList, endEvents).map(eventToOption),
  ];
};

export const initializeFormData = ({ fields, errors }) => {
  const initErrors = fields?.endEventIdentifier
    ? defaultErrors
    : {
        ...defaultErrors,
        endEventIdentifier: !fields?.startEventIdentifier ? [ERRORS.START_EVENT_REQUIRED] : [],
      };
  return {
    fields: {
      ...defaultFields,
      ...fields,
    },
    errors: {
      ...initErrors,
      ...errors,
    },
  };
};

export const validateFields = fields => {
  const newErrors = {};

  if (fields?.name) {
    newErrors.name = DEFAULT_STAGE_NAMES.includes(fields?.name.toLowerCase())
      ? [ERRORS.STAGE_NAME_EXISTS]
      : [];
  }

  if (fields?.startEventIdentifier) {
    newErrors.endEventIdentifier = [];
  } else {
    newErrors.endEventIdentifier = [ERRORS.START_EVENT_REQUIRED];
  }

  if (fields?.endEventIdentifier) {
    newErrors.endEventIdentifier = [];
  }
  return newErrors;
};
