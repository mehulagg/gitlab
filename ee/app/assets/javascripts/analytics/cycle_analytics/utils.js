import { isString } from 'underscore';
import dateFormat from 'dateformat';
import { convertObjectPropsToCamelCase } from '~/lib/utils/common_utils';
import { convertToSnakeCase } from '~/lib/utils/text_utility';
import { newDate, dayAfter, secondsToDays } from '~/lib/utils/datetime_utility';
import { dateFormats } from '../shared/constants';

const EVENT_TYPE_LABEL = 'label';

export const isStartEvent = ev => Boolean(ev) && Boolean(ev.canBeStartEvent) && ev.canBeStartEvent;

export const eventToOption = (obj = null) => {
  if (!obj || (!obj.text && !obj.identifier)) return null;
  const { name: text = '', identifier: value = null } = obj;
  return { text, value };
};

export const getAllowedEndEvents = (events = [], targetIdentifier = null) => {
  if (!targetIdentifier || !events.length) return [];
  const st = events.find(({ identifier }) => identifier === targetIdentifier);
  return st && st.allowedEndEvents ? st.allowedEndEvents : [];
};

export const eventsByIdentifier = (events = [], targetIdentifier = []) => {
  if (!targetIdentifier || !targetIdentifier.length || !events.length) return [];
  return events.filter(({ identifier = '' }) => targetIdentifier.includes(identifier));
};

export const isLabelEvent = (labelEvents = [], ev = null) =>
  Boolean(ev) && labelEvents.length && labelEvents.includes(ev);

export const getLabelEventsIdentifiers = (events = []) =>
  events.filter(ev => ev.type && ev.type === EVENT_TYPE_LABEL).map(i => i.identifier);

/**
 * Checks if the specified stage is in memory or persisted to storage based on the id
 *
 * Default cycle analytics stages are initially stored in memory, when they are first
 * created the id for the stage is the name of the stage in lowercase. This string id
 * is used to fetch stage data (events, median calculation)
 *
 * When either a custom stage is created or an edit is made to a default stage then the
 * default stages get persisted to storage and will have a numeric id. The new numeric
 * id should then be used to access stage data
 *
 * This will be fixed in https://gitlab.com/gitlab-org/gitlab/merge_requests/19278
 */

export const transformRawStages = (stages = []) =>
  stages
    .map(({ id, title, custom = false, ...rest }) => ({
      ...convertObjectPropsToCamelCase(rest, { deep: true }),
      id,
      title,
      slug: custom ? id : convertToSnakeCase(title),
      custom,
      name: title, // editing a stage takes 'name' as a parameter, but the api returns title
    }))
    .sort((a, b) => a.id > b.id);

export const nestQueryStringKeys = (obj = null, targetKey = '') => {
  if (!obj || !isString(targetKey) || !targetKey.length) return {};
  return Object.entries(obj).reduce((prev, [key, value]) => {
    const customKey = `${targetKey}[${key}]`;
    return { ...prev, [customKey]: value };
  }, {});
};

/**
 * Takes the duration data for selected stages, transforms the date values and returns
 * the data in a flattened array
 *
 * The received data is expected to be the following format; One top level object in the array per stage,
 * each potentially having multiple data entries.
 * [
 *   {
 *    slug: 'issue',
 *    selected: true,
 *    data: [
 *      {
 *        'duration_in_seconds': 1234,
 *        'finished_at': '2019-09-02T18:25:43.511Z'
 *      },
 *      ...
 *    ]
 *   },
 *   ...
 * ]
 *
 * The data is then transformed and flattened into the following format;
 * [
 *  {
 *    'duration_in_seconds': 1234,
 *    'finished_at': '2019-09-02'
 *  },
 *  ...
 * ]
 *
 * @param {Array} data - The duration data for selected stages
 * @returns {Array} An array with each item being an object containing the duration_in_seconds and finished_at values for an event
 */
export const flattenDurationChartData = data =>
  data
    .map(stage =>
      stage.data.map(event => {
        const date = new Date(event.finished_at);
        return {
          ...event,
          finished_at: dateFormat(date, dateFormats.isoDate),
        };
      }),
    )
    .flat();

/**
 * Takes the duration data for selected stages, groups the data by day and calculates the total duration
 * per day.
 *
 * The received data is expected to be the following format; One top level object in the array per stage,
 * each potentially having multiple data entries.
 * [
 *   {
 *    slug: 'issue',
 *    selected: true,
 *    data: [
 *      {
 *        'duration_in_seconds': 1234,
 *        'finished_at': '2019-09-02T18:25:43.511Z'
 *      },
 *      ...
 *    ]
 *   },
 *   ...
 * ]
 *
 * The data is then computed and transformed into a format that can be passed to the chart:
 * [
 *  ['2019-09-02', 7, '2019-09-02'],
 *  ['2019-09-03', 10, '2019-09-03'],
 *  ['2019-09-04', 8, '2019-09-04'],
 *  ...
 * ]
 *
 * In the data above, each array i represents a point in the scatterplot with the following data:
 * i[0] = date, displayed on x axis
 * i[1] = metric, displayed on y axis
 * i[2] = date, used in the tooltip
 *
 * @param {Array} data - The duration data for selected stages
 * @param {Date} startDate - The globally selected cycle analytics start date
 * @param {Date} endDate - The globally selected cycle analytics stendart date
 * @returns {Array} An array with each item being another arry of three items (plottable date, computed total, tooltip display date)
 */
export const getDurationChartData = (data, startDate, endDate) => {
  const flattenedData = flattenDurationChartData(data);
  const eventData = [];

  for (
    let currentDate = newDate(startDate);
    currentDate <= endDate;
    currentDate = dayAfter(currentDate)
  ) {
    const currentISODate = dateFormat(newDate(currentDate), dateFormats.isoDate);
    const valuesForDay = flattenedData.filter(object => object.finished_at === currentISODate);
    const summedData = valuesForDay.reduce((total, value) => total + value.duration_in_seconds, 0);
    const summedDataInDays = secondsToDays(summedData);

    if (summedDataInDays) eventData.push([currentISODate, summedDataInDays, currentISODate]);
  }

  return eventData;
};
