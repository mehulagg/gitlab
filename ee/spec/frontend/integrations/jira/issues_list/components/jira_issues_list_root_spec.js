import { shallowMount } from '@vue/test-utils';
import MockAdapter from 'axios-mock-adapter';

import JiraIssuesListRoot from 'ee/integrations/jira/issues_list/components/jira_issues_list_root.vue';
import waitForPromises from 'helpers/wait_for_promises';

import createFlash from '~/flash';
import IssuableList from '~/issuable_list/components/issuable_list_root.vue';
import { IssuableStates, IssuableListTabs, AvailableSortOptions } from '~/issuable_list/constants';
import axios from '~/lib/utils/axios_utils';
import { convertObjectPropsToCamelCase } from '~/lib/utils/common_utils';
import httpStatus from '~/lib/utils/http_status';

import { mockProvide, mockJiraIssues } from '../mock_data';

jest.mock('~/flash');
jest.mock('~/issuable_list/constants', () => ({
  DEFAULT_PAGE_SIZE: 2,
  IssuableStates: jest.requireActual('~/issuable_list/constants').IssuableStates,
  IssuableListTabs: jest.requireActual('~/issuable_list/constants').IssuableListTabs,
  AvailableSortOptions: jest.requireActual('~/issuable_list/constants').AvailableSortOptions,
}));

const resolvedValue = {
  headers: {
    'x-page': 1,
    'x-total': 3,
  },
  data: mockJiraIssues,
};

describe('JiraIssuesListRoot', () => {
  let wrapper;
  let mock;

  const findIssuableList = () => wrapper.find(IssuableList);

  const createComponent = ({ provide = mockProvide, initialFilterParams = {}, data = {} } = {}) => {
    wrapper = shallowMount(JiraIssuesListRoot, {
      propsData: {
        initialFilterParams,
      },
      provide,
      data() {
        return data;
      },
    });
  };

  beforeEach(() => {
    mock = new MockAdapter(axios);
  });

  afterEach(() => {
    wrapper.destroy();
    mock.restore();
  });

  describe('computed', () => {
    describe('urlParams', () => {
      it('returns object containing `state`, `page`, `sort` and `search` properties', () => {
        createComponent({
          data: {
            currentState: 'closed',
            currentPage: 2,
            sortedBy: 'created_asc',
            filterParams: {
              search: 'foo',
            },
          },
        });

        expect(wrapper.vm.urlParams).toMatchObject({
          state: 'closed',
          page: 2,
          sort: 'created_asc',
          search: 'foo',
        });
      });
    });
  });

  describe('methods', () => {
    describe('fetchIssuesBy', () => {
      it('sets provided prop value for given prop name and calls `fetchIssues`', () => {
        jest.spyOn(axios, 'get');
        createComponent();

        wrapper.vm.fetchIssuesBy('currentPage', 2);

        expect(wrapper.vm.currentPage).toBe(2);
        expect(axios.get).toHaveBeenCalled();
      });
    });
  });

  describe('template', () => {
    it('renders issuable-list component', async () => {
      createComponent({
        data: {
          filterParams: {
            search: 'foo',
          },
        },
      });

      await waitForPromises();

      const issuableList = findIssuableList();
      expect(issuableList.exists()).toBe(true);
      expect(issuableList.props()).toMatchObject({
        namespace: mockProvide.projectFullPath,
        tabs: IssuableListTabs,
        currentTab: 'opened',
        searchInputPlaceholder: 'Search Jira issues',
        searchTokens: [],
        sortOptions: AvailableSortOptions,
        initialFilterValue: [
          {
            type: 'filtered-search-term',
            value: {
              data: 'foo',
            },
          },
        ],
        initialSortBy: 'created_desc',
        issuables: [],
        issuablesLoading: false,
        showPaginationControls: wrapper.vm.showPaginationControls,
        defaultPageSize: 2, // mocked value in tests
        totalItems: 0,
        currentPage: 1,
        previousPage: 0,
        nextPage: 2,
        urlParams: wrapper.vm.urlParams,
        recentSearchesStorageKey: 'jira_issues',
        enableLabelPermalinks: true,
      });
    });

    describe('issuable-list events', () => {
      beforeEach(async () => {
        jest.spyOn(axios, 'get');
        createComponent();
        await waitForPromises();
      });

      it('click-tab event changes currentState value and calls fetchIssues via `fetchIssuesBy`', () => {
        findIssuableList().vm.$emit('click-tab', 'closed');

        expect(wrapper.vm.currentState).toBe('closed');
        expect(axios.get).toHaveBeenCalled();
      });

      it('page-change event changes currentPage value and calls fetchIssues via `fetchIssuesBy`', () => {
        findIssuableList().vm.$emit('page-change', 2);

        expect(wrapper.vm.currentPage).toBe(2);
        expect(axios.get).toHaveBeenCalled();
      });

      it('sort event changes sortedBy value and calls fetchIssues via `fetchIssuesBy`', () => {
        findIssuableList().vm.$emit('sort', 'updated_asc');

        expect(wrapper.vm.sortedBy).toBe('updated_asc');
        expect(axios.get).toHaveBeenCalled();
      });

      it('filter event sets `filterParams` value and calls fetchIssues', () => {
        findIssuableList().vm.$emit('filter', [
          {
            type: 'filtered-search-term',
            value: {
              data: 'foo',
            },
          },
        ]);

        expect(wrapper.vm.filterParams).toEqual({
          search: 'foo',
        });
        expect(axios.get).toHaveBeenCalled();
      });
    });

    describe('pagination', () => {
      it.each`
        scenario               | issuesListLoadFailed | issues            | shouldShowPaginationControls
        ${'fails'}             | ${true}              | ${[]}             | ${false}
        ${'returns no issues'} | ${false}             | ${[]}             | ${false}
        ${`returns issues`}    | ${false}             | ${mockJiraIssues} | ${true}
      `(
        'sets `showPaginationControls` prop to $shouldShowPaginationControls when request $scenario',
        async ({ issuesListLoadFailed, issues, shouldShowPaginationControls }) => {
          jest.spyOn(axios, 'get');
          mock
            .onGet(mockProvide.issuesFetchPath)
            .replyOnce(
              issuesListLoadFailed ? httpStatus.INTERNAL_SERVER_ERROR : httpStatus.OK,
              issues,
              {
                'x-page': 1,
                'x-total': 3,
              },
            );

          createComponent();

          await waitForPromises();

          expect(findIssuableList().props('showPaginationControls')).toBe(
            shouldShowPaginationControls,
          );
        },
      );
    });
  });

  describe('on mount', () => {
    describe('while loading', () => {
      it('sets issuesListLoading to `true`', () => {
        jest.spyOn(axios, 'get').mockResolvedValue(new Promise(() => {}));

        createComponent();

        const issuableList = findIssuableList();
        expect(issuableList.props('issuablesLoading')).toBe(true);
      });

      it('calls `axios.get` with `issuesFetchPath` and query params', () => {
        jest.spyOn(axios, 'get');
        createComponent();

        expect(axios.get).toHaveBeenCalledWith(
          mockProvide.issuesFetchPath,
          expect.objectContaining({
            params: {
              with_labels_details: true,
              page: wrapper.vm.currentPage,
              per_page: wrapper.vm.$options.defaultPageSize,
              state: wrapper.vm.currentState,
              sort: wrapper.vm.sortedBy,
              search: wrapper.vm.filterParams.search,
            },
          }),
        );
      });
    });

    describe('when request succeeds', () => {
      beforeEach(async () => {
        jest.spyOn(axios, 'get').mockResolvedValue(resolvedValue);

        createComponent();
        await waitForPromises();
      });

      it('sets `currentPage` and `totalIssues` from response headers and `issues` & `issuesCount` from response body when request is successful', async () => {
        const firstIssue = convertObjectPropsToCamelCase(mockJiraIssues[0], { deep: true });

        expect(wrapper.vm.currentPage).toBe(resolvedValue.headers['x-page']);
        expect(wrapper.vm.totalIssues).toBe(resolvedValue.headers['x-total']);
        expect(wrapper.vm.issues[0]).toEqual({
          ...firstIssue,
          id: 31596,
          author: {
            ...firstIssue.author,
            id: 0,
          },
        });
        expect(wrapper.vm.issuesCount[IssuableStates.Opened]).toBe(3);
      });

      it('sets issuesListLoading to `false`', () => {
        const issuableList = findIssuableList();
        expect(issuableList.props('issuablesLoading')).toBe(false);
      });
    });

    describe('when request fails', () => {
      it.each`
        APIErrorMessage | expectedRenderedErrorMessage
        ${'API error'}  | ${'An error occurred while loading issues'}
        ${undefined}    | ${'An error occurred while loading issues'}
      `(
        'calls `createFlash` with "$expectedRenderedErrorMessage" when API responds with "$APIErrorMessage"',
        async ({ APIErrorMessage, expectedRenderedErrorMessage }) => {
          jest.spyOn(axios, 'get');
          mock
            .onGet(mockProvide.issuesFetchPath)
            .replyOnce(httpStatus.INTERNAL_SERVER_ERROR, { errors: [APIErrorMessage] });

          createComponent();

          await waitForPromises();

          expect(createFlash).toHaveBeenCalledWith({
            message: expectedRenderedErrorMessage,
            captureError: true,
            error: expect.any(Object),
          });
        },
      );
    });
  });
});
