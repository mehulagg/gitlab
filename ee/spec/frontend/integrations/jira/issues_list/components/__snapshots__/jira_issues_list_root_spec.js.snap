// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JiraIssuesListRoot when request succeeds renders issuable-list component with correct props 1`] = `
Object {
  "currentPage": 1,
  "currentTab": "opened",
  "defaultPageSize": 2,
  "enableLabelPermalinks": true,
  "hasNextPage": false,
  "hasPreviousPage": false,
  "initialFilterValue": Array [
    Object {
      "type": "filtered-search-term",
      "value": Object {
        "data": "",
      },
    },
  ],
  "initialSortBy": "created_desc",
  "isManualOrdering": false,
  "issuableSymbol": "#",
  "issuables": Array [
    Object {
      "assignees": Array [
        Object {
          "avatarUrl": null,
          "name": "Kushal Pandya",
          "webUrl": "https://gitlab-jira.atlassian.net/people/1920938475",
        },
      ],
      "author": Object {
        "avatarUrl": null,
        "name": "jhope",
        "webUrl": "https://gitlab-jira.atlassian.net/people/5e32f803e127810e82875bc1",
      },
      "closedAt": null,
      "createdAt": "2020-03-19T14:31:51.281Z",
      "externalTracker": "jira",
      "gitlabWebUrl": "",
      "id": 31596,
      "labels": Array [
        Object {
          "color": "#0052CC",
          "name": "backend",
          "textColor": "#FFFFFF",
          "title": "backend",
        },
      ],
      "projectId": 1,
      "references": Object {
        "relative": "IG-31596",
      },
      "status": "Selected for Development",
      "title": "Eius fuga voluptates.",
      "updatedAt": "2020-10-20T07:01:45.865Z",
      "webUrl": "https://gitlab-jira.atlassian.net/browse/IG-31596",
    },
    Object {
      "assignees": Array [],
      "author": Object {
        "avatarUrl": null,
        "name": "Gabe Weaver",
        "webUrl": "https://gitlab-jira.atlassian.net/people/5e320a31fe03e20c9d1dccde",
      },
      "closedAt": null,
      "createdAt": "2020-03-19T14:31:50.677Z",
      "externalTracker": "jira",
      "gitlabWebUrl": "",
      "id": 31595,
      "labels": Array [],
      "projectId": 1,
      "references": Object {
        "relative": "IG-31595",
      },
      "status": "Backlog",
      "title": "Hic sit sint ducimus ea et sint.",
      "updatedAt": "2020-03-19T14:31:50.677Z",
      "webUrl": "https://gitlab-jira.atlassian.net/browse/IG-31595",
    },
    Object {
      "assignees": Array [],
      "author": Object {
        "avatarUrl": null,
        "name": "Gabe Weaver",
        "webUrl": "https://gitlab-jira.atlassian.net/people/5e320a31fe03e20c9d1dccde",
      },
      "closedAt": null,
      "createdAt": "2020-03-19T14:31:50.012Z",
      "externalTracker": "jira",
      "gitlabWebUrl": "",
      "id": 31594,
      "labels": Array [],
      "projectId": 1,
      "references": Object {
        "relative": "IG-31594",
      },
      "status": "Backlog",
      "title": "Alias ut modi est labore.",
      "updatedAt": "2020-03-19T14:31:50.012Z",
      "webUrl": "https://gitlab-jira.atlassian.net/browse/IG-31594",
    },
  ],
  "issuablesLoading": false,
  "labelFilterParam": "labels",
  "namespace": "gitlab-org/gitlab-test",
  "nextPage": 2,
  "previousPage": 0,
  "recentSearchesStorageKey": "jira_issues",
  "searchInputPlaceholder": "Search Jira issues",
  "searchTokens": Array [],
  "showBulkEditSidebar": false,
  "showPaginationControls": true,
  "sortOptions": Array [
    Object {
      "id": 1,
      "sortDirection": Object {
        "ascending": "created_asc",
        "descending": "created_desc",
      },
      "title": "Created date",
    },
    Object {
      "id": 2,
      "sortDirection": Object {
        "ascending": "updated_asc",
        "descending": "updated_desc",
      },
      "title": "Last updated",
    },
  ],
  "tabCounts": null,
  "tabs": Array [
    Object {
      "id": "state-opened",
      "name": "opened",
      "title": "Open",
      "titleTooltip": "Filter by issues that are currently opened.",
    },
    Object {
      "id": "state-closed",
      "name": "closed",
      "title": "Closed",
      "titleTooltip": "Filter by issues that are currently closed.",
    },
    Object {
      "id": "state-all",
      "name": "all",
      "title": "All",
      "titleTooltip": "Show all issues.",
    },
  ],
  "totalItems": 3,
  "urlParams": Object {
    "labels[]": undefined,
    "search": undefined,
  },
  "useKeysetPagination": false,
}
`;
